File: year_2022/day_10/day10.go
package main

import (
	"aoc/tools"
	"fmt"
	"strconv"
	"strings"
)

func solve(input string) (int, string) {
	var (
		instructions                              = tools.ReadLines(input)
		register                                  = 1
		cycle, index, skip, addValue, part1, hPos int
		part2                                     string
	)

	for index < len(instructions) {
		cycle++
		hPos++

		if skip == 1 {
			register += addValue
		}

		if register <= hPos && hPos <= register+2 {
			part2 += "█"
		} else {
			part2 += "."
		}

		if cycle == 20 || (cycle+20)%40 == 0 {
			part1 += cycle * register
		} else if cycle%40 == 0 {
			part2 += "\n"
			hPos = 0
		}

		if skip == 2 {
			skip--

			continue
		}

		line := instructions[index]

		if line != "noop" {
			addValue, _ = strconv.Atoi(strings.Fields(line)[1])
			skip = 2
		} else {
			addValue = 0
		}

		index++
	}

	return part1, part2
}

func main() {
	data := [2]string{"test10.txt", tools.GetData(2022, 10)}

	for _, input := range data {
		part1, part2 := solve(input)
		fmt.Println(part1)
		fmt.Println(part2)
	}
}

File: year_2022/day_10/day10.py
from tools import loader, parsers


class CRT:
    def __init__(self, data: list[str]) -> None:
        self.next_line = iter(data)
        self.image = ''
        self.out = {}
        self.run()

    def run(self) -> None:
        x = 1
        tick = timer = h_pos = value = 0
        cmd = ''

        while True:
            if tick % 40 == 0:
                self.image += '\n'
                h_pos = 0
            elif tick == 20 or (tick + 20) % 40 == 0:
                self.out[tick] = x
            tick += 1
            h_pos += 1

            if timer > 1:
                timer -= 1
            else:
                if cmd == 'addx' and timer == 1:
                    x += value
                try:
                    line = next(self.next_line).split()
                except StopIteration:
                    break

                cmd = line[0]
                value = int(line[1]) if len(line) > 1 else None
                timer = 2 if cmd == 'addx' else 1

            self.image += '█' if x <= h_pos <= x + 2 else '⠀'

    def part_1(self) -> int:
        """test part 1:
        >>> print(CRT(parsers.lines('test10.txt')).part_1())
        13140
        """
        return sum([i * self.out[i] for i in self.out])

    def part_2(self) -> str:
        """test part 1:
        >>> print(CRT(parsers.lines('test10.txt')).part_2())
        ██⠀⠀██⠀⠀██⠀⠀██⠀⠀██⠀⠀██⠀⠀██⠀⠀██⠀⠀██⠀⠀██⠀⠀
        ███⠀⠀⠀███⠀⠀⠀███⠀⠀⠀███⠀⠀⠀███⠀⠀⠀███⠀⠀⠀███⠀
        ████⠀⠀⠀⠀████⠀⠀⠀⠀████⠀⠀⠀⠀████⠀⠀⠀⠀████⠀⠀⠀⠀
        █████⠀⠀⠀⠀⠀█████⠀⠀⠀⠀⠀█████⠀⠀⠀⠀⠀█████⠀⠀⠀⠀⠀
        ██████⠀⠀⠀⠀⠀⠀██████⠀⠀⠀⠀⠀⠀██████⠀⠀⠀⠀⠀⠀████
        ███████⠀⠀⠀⠀⠀⠀⠀███████⠀⠀⠀⠀⠀⠀⠀███████⠀⠀⠀⠀⠀
        """
        return self.image.strip()


print(CRT(parsers.lines(loader.get())).part_1())  # 15680
print(CRT(parsers.lines(loader.get())).part_2())  # ZFBFHGUP
